File: submissions/the-randoms/src/server-custom-llm/node/conversation_store.js
/**
 * In-memory conversation store for the Custom LLM Server.
 *
 * Stores conversations keyed by appId:userId:channel. Includes automatic
 * trimming when conversations exceed MAX_MESSAGES and periodic cleanup
 * of stale conversations.
 */

const logger = {
  info: (message) => console.log(`INFO: ${message}`),
  debug: (message) => console.log(`DEBUG: ${message}`),
  error: (message, error) => console.error(`ERROR: ${message}`, error),
};

// Limits
const MAX_MESSAGES = 100;
const TARGET_MESSAGES = 75;
const CLEANUP_INTERVAL = 3600 * 1000; // 1 hour in ms
const MAX_AGE = 86400 * 1000; // 24 hours in ms

// In-memory store: key -> { messages: [...], lastUpdated: timestamp }
const conversations = {};

function getConversationKey(appId, userId, channel) {
  return `${appId}:${userId}:${channel}`;
}

function getOrCreateConversation(appId, userId, channel) {
  const key = getConversationKey(appId, userId, channel);
  if (!conversations[key]) {
    logger.info(`Creating new conversation: ${key}`);
    conversations[key] = {
      messages: [],
      lastUpdated: Date.now(),
    };
  }
  return conversations[key];
}

/**
 * Append a message to the conversation and trim if necessary.
 *
 * message should be an object with at least { role, content }.
 * Tool messages should also have tool_call_id and name.
 * Assistant tool_call messages should have tool_calls.
 */
function saveMessage(appId, userId, channel, message) {
  const conv = getOrCreateConversation(appId, userId, channel);
  const msg = { ...message, timestamp: Date.now() };
  conv.messages.push(msg);
  conv.lastUpdated = Date.now();

  if (conv.messages.length > MAX_MESSAGES) {
    trimConversation(conv);
  }

  logger.debug(
    `Saved ${message.role} message, total=${conv.messages.length} [${appId}:${userId}:${channel}]`
  );
}

/**
 * Return a copy of the conversation messages.
 */
function getMessages(appId, userId, channel) {
  const conv = getOrCreateConversation(appId, userId, channel);
  return [...conv.messages];
}

/**
 * Trim a conversation to TARGET_MESSAGES, preserving:
 * 1. All system messages
 * 2. Tool call pairs (assistant with tool_calls + tool response)
 * 3. Most recent non-system messages
 */
function trimConversation(conv) {
  const messages = conv.messages;
  const systemMsgs = messages.filter((m) => m.role === 'system');
  const nonSystem = messages.filter((m) => m.role !== 'system');

  // Keep the most recent TARGET_MESSAGES non-system messages
  let kept = nonSystem.slice(-TARGET_MESSAGES);

  // Collect tool_call IDs that are in kept messages
  const toolCallIds = new Set();
  for (const m of kept) {
    if (m.role === 'tool' && m.tool_call_id) {
      toolCallIds.add(m.tool_call_id);
    }
    if (m.role === 'assistant' && m.tool_calls) {
      for (const tc of m.tool_calls) {
        if (tc.id) toolCallIds.add(tc.id);
      }
    }
  }

  // Add orphaned pair messages
  for (const m of nonSystem) {
    if (kept.includes(m)) continue;
    if (m.role === 'assistant' && m.tool_calls) {
      for (const tc of m.tool_calls) {
        if (tc.id && toolCallIds.has(tc.id)) {
          kept.unshift(m);
          break;
        }
      }
    } else if (m.role === 'tool' && toolCallIds.has(m.tool_call_id)) {
      kept.unshift(m);
    }
  }

  // Sort by timestamp
  kept.sort((a, b) => (a.timestamp || 0) - (b.timestamp || 0));

  conv.messages = [...systemMsgs, ...kept];
  logger.debug(
    `Trimmed conversation: ${messages.length} -> ${conv.messages.length}`
  );
}

/**
 * Remove conversations older than MAX_AGE.
 */
function cleanupOldConversations() {
  const now = Date.now();
  let removed = 0;
  for (const key of Object.keys(conversations)) {
    if (now - conversations[key].lastUpdated > MAX_AGE) {
      delete conversations[key];
      removed++;
    }
  }
  if (removed) {
    logger.info(`Cleaned up ${removed} old conversation(s)`);
  }
}

// Start periodic cleanup
const cleanupTimer = setInterval(() => {
  try {
    cleanupOldConversations();
  } catch (e) {
    logger.error('Cleanup error:', e);
  }
}, CLEANUP_INTERVAL);

// Don't let the timer prevent process exit
if (cleanupTimer.unref) {
  cleanupTimer.unref();
}

logger.info(
  `Conversation store initialized (max=${MAX_MESSAGES}, target=${TARGET_MESSAGES}, cleanup every ${CLEANUP_INTERVAL / 1000}s)`
);

module.exports = {
  getOrCreateConversation,
  saveMessage,
  getMessages,
  cleanupOldConversations,
};

File: submissions/the-randoms/src/server-custom-llm/python/conversation_store.py
"""
In-memory conversation store for the Custom LLM Server.

Stores conversations keyed by appId:userId:channel. Includes automatic
trimming when conversations exceed MAX_MESSAGES and periodic cleanup
of stale conversations.
"""

import time
import threading
import logging
from typing import Dict, List, Optional

logger = logging.getLogger(__name__)

# Limits
MAX_MESSAGES = 100
TARGET_MESSAGES = 75
CLEANUP_INTERVAL = 3600  # 1 hour
MAX_AGE = 86400  # 24 hours

# In-memory store: key -> {"messages": [...], "last_updated": float}
_conversations: Dict[str, dict] = {}
_lock = threading.Lock()


def get_conversation_key(app_id: str, user_id: str, channel: str) -> str:
    return f"{app_id}:{user_id}:{channel}"


def get_or_create_conversation(app_id: str, user_id: str, channel: str) -> dict:
    key = get_conversation_key(app_id, user_id, channel)
    with _lock:
        if key not in _conversations:
            logger.info(f"Creating new conversation: {key}")
            _conversations[key] = {
                "messages": [],
                "last_updated": time.time(),
            }
        return _conversations[key]


def save_message(
    app_id: str, user_id: str, channel: str, message: dict
) -> None:
    """
    Append a message to the conversation and trim if necessary.

    message should be a dict with at least 'role' and 'content'.
    Tool messages should also have 'tool_call_id' and 'name'.
    Assistant tool_call messages should have 'tool_calls'.
    """
    conv = get_or_create_conversation(app_id, user_id, channel)
    msg = {**message, "timestamp": time.time()}
    with _lock:
        conv["messages"].append(msg)
        conv["last_updated"] = time.time()
        if len(conv["messages"]) > MAX_MESSAGES:
            _trim_conversation(conv)

    logger.debug(
        f"Saved {message.get('role')} message, "
        f"total={len(conv['messages'])} [{app_id}:{user_id}:{channel}]"
    )


def get_messages(app_id: str, user_id: str, channel: str) -> List[dict]:
    """Return a copy of the conversation messages."""
    conv = get_or_create_conversation(app_id, user_id, channel)
    with _lock:
        return list(conv["messages"])


def _trim_conversation(conv: dict) -> None:
    """
    Trim a conversation to TARGET_MESSAGES, preserving:
    1. All system messages
    2. Tool call pairs (assistant with tool_calls + tool response)
    3. Most recent non-system messages
    """
    messages = conv["messages"]
    system_msgs = [m for m in messages if m.get("role") == "system"]
    non_system = [m for m in messages if m.get("role") != "system"]

    # Keep the most recent TARGET_MESSAGES non-system messages
    kept = non_system[-TARGET_MESSAGES:]

    # Ensure tool message pairs are intact
    tool_call_ids_in_kept = set()
    for m in kept:
        if m.get("role") == "tool" and m.get("tool_call_id"):
            tool_call_ids_in_kept.add(m["tool_call_id"])
        if m.get("role") == "assistant" and m.get("tool_calls"):
            for tc in m["tool_calls"]:
                tc_id = tc.get("id") if isinstance(tc, dict) else None
                if tc_id:
                    tool_call_ids_in_kept.add(tc_id)

    # Check for orphaned tool messages and add their pairs
    for m in non_system:
        if m in kept:
            continue
        if m.get("role") == "assistant" and m.get("tool_calls"):
            for tc in m["tool_calls"]:
                tc_id = tc.get("id") if isinstance(tc, dict) else None
                if tc_id and tc_id in tool_call_ids_in_kept:
                    kept.insert(0, m)
                    break
        elif m.get("role") == "tool" and m.get("tool_call_id") in tool_call_ids_in_kept:
            kept.insert(0, m)

    # Sort by timestamp
    kept.sort(key=lambda m: m.get("timestamp", 0))

    conv["messages"] = system_msgs + kept
    logger.debug(
        f"Trimmed conversation: {len(messages)} -> {len(conv['messages'])}"
    )


def cleanup_old_conversations() -> None:
    """Remove conversations older than MAX_AGE."""
    now = time.time()
    removed = 0
    with _lock:
        keys_to_remove = [
            key
            for key, conv in _conversations.items()
            if now - conv["last_updated"] > MAX_AGE
        ]
        for key in keys_to_remove:
            del _conversations[key]
            removed += 1

    if removed:
        logger.info(f"Cleaned up {removed} old conversation(s)")


def _cleanup_loop() -> None:
    """Background thread that runs cleanup periodically."""
    while True:
        time.sleep(CLEANUP_INTERVAL)
        try:
            cleanup_old_conversations()
        except Exception as e:
            logger.error(f"Cleanup error: {e}")


# Start the background cleanup thread
_cleanup_thread = threading.Thread(target=_cleanup_loop, daemon=True)
_cleanup_thread.start()
logger.info(
    f"Conversation store initialized (max={MAX_MESSAGES}, "
    f"target={TARGET_MESSAGES}, cleanup every {CLEANUP_INTERVAL}s)"
)
